sdks/python/apache_beam/dataframe/convert.py

```diff
@@ -67,6 +68,9 @@ def to_dataframe(
       expressions.PlaceholderExpression(proxy, pcoll))
 
 
+TO_PCOLLECTION_CACHE = {}
```

Comment:
Note this uses a global cache which can be problematic in testing. I don't think we need to worry about pcollections being shared across pipelines though since the dataframe expressions should have a reference to the pipeline through the to_dataframe roots.

Reply:
I'm actually more worried about global caches in production than testing--the expressions themselves should not collide between pipelines. 

Maybe we could use a weakref.WeakKeyDictionary.

Reply:
Yeah agreed it could be a problem in production as well. `WeakKeyDictionary` is an interesting idea, I hadn't seen the `weakref` library before. I'll give that a shot.

Reply:
Updated this to use `WeakValueDictionary` for the caches, PTAL